tests/diffs/test_diff.py

```diff
@@ -0,0 +1,78 @@
+from kopf.structs.diffs import diff
+
+
+def test_scalars_equal():
+    a = 100
+    b = 100
+    d = diff(a, b)
+    assert d == ()
+
+
+def test_scalars_unequal():
+    a = 100
+    b = 200
+    d = diff(a, b)
+    assert d == (('change', (), 100, 200),)
+
+
+def test_strings_equal():
+    a = 'hello'
+    b = 'hello'
+    d = diff(a, b)
+    assert d == ()
+
+
+def test_strings_unequal():
+    a = 'hello'
+    b = 'world'
+    d = diff(a, b)
+    assert d == (('change', (), 'hello', 'world'),)
+
+
+def test_lists_equal():
+    a = [100, 200, 300]
+    b = [100, 200, 300]
+    d = diff(a, b)
+    assert d == ()
+
+
+def test_lists_unequal():
+    a = [100, 200, 300]
+    b = [100, 666, 300]
+    d = diff(a, b)
+    assert d == (('change', (), [100, 200, 300], [100, 666, 300]),)
+
+
+def test_dicts_equal():
+    a = {'hello': 'world', 'key': 'val'}
+    b = {'key': 'val', 'hello': 'world'}
+    d = diff(a, b)
+    assert d == ()
+
+
+def test_dicts_with_keys_added():
+    a = {'hello': 'world'}
+    b = {'hello': 'world', 'key': 'val'}
+    d = diff(a, b)
+    assert d == (('add', ('key',), None, 'val'),)
+
+
+def test_dicts_with_keys_removed():
+    a = {'hello': 'world', 'key': 'val'}
+    b = {'hello': 'world'}
+    d = diff(a, b)
+    assert d == (('remove', ('key',), 'val', None),)
+
+
+def test_dicts_with_keys_changed():
+    a = {'hello': 'world', 'key': 'old'}
+    b = {'hello': 'world', 'key': 'new'}
+    d = diff(a, b)
+    assert d == (('change', ('key',), 'old', 'new'),)
+
+
+def test_dicts_with_subkeys_changed():
+    a = {'main': {'hello': 'world', 'key': 'old'}}
+    b = {'main': {'hello': 'world', 'key': 'new'}}
+    d = diff(a, b)
+    assert d == (('change', ('main', 'key'), 'old', 'new'),)
```

Comment:
I would like to have a test with a real usecase with Dicts that the reactor is using; which would get closer to 'test as documentation'

```suggestion
def test_dicts_sample_kopf_usecase_update():
    body_before_update = {'metadata' : {'annotations' : 'abcdef'}}
    body_after_update = {'metadata' : {'annotations' : 'abcdef'}}
```

Reply:
@parking52 Good idea. 

Though, the `resourceVersion` will never be in the diffs, as this and other system fields are filtered out of the last-seen state calculation (see [`kopf.structs.lastseen.get_state()`](https://github.com/zalando-incubator/kopf/blob/fa4d77244dd1dd0e95f89b0f553173649d11db04/kopf/structs/lastseen.py#L30-L45)) — for the reason that these fields change even when we do not care about these changes. 

Only all of the the `spec` fields & some (non-filtered) `metainfo` fields, such as `metainfo.labels`, `metainfo.annotations` (except few specific annotations) will be taken into account, diffs calculated, and the update-handlers called.

Maybe, you can implement it for all of the mentioned above? I.e.:

* metainfo.labels.some-label
* metainfo.annotations.some-annotation
* spec.field (a string)
* spec.items (a list of strings)
* spec.modes (a dict of strings=>strings).

These are the main real-world examples happening in the operators and expected in `@kopf.on.field()`. 

See also: [`examples/obj.yaml`](https://github.com/zalando-incubator/kopf/blob/fa4d77244dd1dd0e95f89b0f553173649d11db04/examples/obj.yaml), which I use(d) for the local runs and manual testing.